File: backend/app/utils/formatacao.py
"""
Formatação de documentos.

Regra do projeto: no banco, CPF e MASP ficam SEMPRE guardados só com os
dígitos (sem ponto, traço ou espaço). A formatação é aplicada apenas na
hora de mostrar o dado — em tela ou no relatório. Isso evita o problema
clássico de o mesmo CPF entrar duas vezes, uma com máscara e outra sem,
e passar pela verificação de duplicidade.
"""

import re


def apenas_digitos(valor: str | None) -> str | None:
    """Remove tudo que não é número. Devolve None quando sobra nada."""
    if valor is None:
        return None

    digitos = re.sub(r"\D", "", valor)

    return digitos or None


def formatar_cpf(cpf: str | None) -> str | None:
    """000.000.000-00 — devolve o valor original se não tiver 11 dígitos."""
    digitos = apenas_digitos(cpf)

    if not digitos:
        return None

    if len(digitos) != 11:
        return cpf

    return (
        f"{digitos[:3]}.{digitos[3:6]}.{digitos[6:9]}-{digitos[9:]}"
    )


def formatar_masp(masp: str | None) -> str | None:
    """
    Formato oficial do MASP (Minas Gerais): 7 dígitos de matrícula mais
    1 dígito verificador, separados por hífen — ex.: 1234567-8.

    MASPs com quantidade diferente de dígitos são devolvidos como estão,
    para não desfigurar matrículas antigas ou provisórias.
    """
    digitos = apenas_digitos(masp)

    if not digitos:
        return None

    if len(digitos) != 8:
        return masp

    return f"{digitos[:7]}-{digitos[7:]}"

File: frontend/src/utils/formatacao.js
/**
 * Formatação e máscara de documentos.
 *
 * Regra do projeto: o backend guarda CPF e MASP só com os dígitos.
 * A máscara existe apenas na tela — tanto para digitar quanto para
 * exibir. Toda vez que um valor é enviado para a API, ele passa por
 * `apenasDigitos`.
 */

export function apenasDigitos(valor) {
  return String(valor ?? "").replace(/\D/g, "");
}

/** Exibe o CPF como 000.000.000-00. */
export function formatarCpf(valor) {
  const digitos = apenasDigitos(valor);

  if (digitos.length !== 11) {
    return valor || "";
  }

  return digitos.replace(
    /(\d{3})(\d{3})(\d{3})(\d{2})/,
    "$1.$2.$3-$4"
  );
}

/**
 * Exibe o MASP no formato oficial de Minas Gerais: 7 dígitos de
 * matrícula + 1 dígito verificador (ex.: 1234567-8). MASPs com outra
 * quantidade de dígitos são mostrados como estão.
 */
export function formatarMasp(valor) {
  const digitos = apenasDigitos(valor);

  if (digitos.length !== 8) {
    return valor || "";
  }

  return `${digitos.slice(0, 7)}-${digitos.slice(7)}`;
}

/**
 * Máscara progressiva de CPF, aplicada enquanto a pessoa digita:
 * "144" → "144", "1441463" → "144.146.3", e assim por diante.
 */
export function mascararCpf(valor) {
  const digitos = apenasDigitos(valor).slice(0, 11);

  return digitos
    .replace(/(\d{3})(\d)/, "$1.$2")
    .replace(/(\d{3})\.(\d{3})(\d)/, "$1.$2.$3")
    .replace(/(\d{3})\.(\d{3})\.(\d{3})(\d)/, "$1.$2.$3-$4");
}

/** Máscara progressiva de MASP (1234567-8). */
export function mascararMasp(valor) {
  const digitos = apenasDigitos(valor).slice(0, 8);

  return digitos.replace(/(\d{7})(\d)/, "$1-$2");
}

/** Máscara de telefone: (31) 99999-8888 ou (31) 3999-8888. */
export function mascararTelefone(valor) {
  const digitos = apenasDigitos(valor).slice(0, 11);

  if (digitos.length <= 10) {
    return digitos
      .replace(/(\d{2})(\d)/, "($1) $2")
      .replace(/(\d{4})(\d)/, "$1-$2");
  }

  return digitos
    .replace(/(\d{2})(\d)/, "($1) $2")
    .replace(/(\d{5})(\d)/, "$1-$2");
}

/**
 * Aplica a máscara certa dependendo do que a pessoa está digitando na
 * busca: só números com cara de CPF viram CPF, com cara de MASP viram
 * MASP, e qualquer coisa com letra é tratada como nome e fica intacta.
 */
export function mascararDocumentoOuNome(valor) {
  const texto = String(valor ?? "");

  if (/[^\d.\-\s]/.test(texto)) {
    return texto;
  }

  const digitos = apenasDigitos(texto);

  if (digitos.length > 8) {
    return mascararCpf(texto);
  }

  return mascararMasp(texto);
}

/**
 * Linha "CPF: ... · MASP: ..." usada nas listas. Prefere os campos já
 * formatados pelo backend e, se não vierem, formata aqui mesmo.
 */
export function resumoDocumentos(cidadao) {
  if (!cidadao) return "";

  const cpf = cidadao.cpf_formatado || formatarCpf(cidadao.cpf);
  const masp = cidadao.masp_formatado || formatarMasp(cidadao.masp);

  return [
    cpf ? `CPF: ${cpf}` : null,
    masp ? `MASP: ${masp}` : null,
  ]
    .filter(Boolean)
    .join(" · ");
}
